drivers/rtc/rtc_nxp_irtc.c

```diff
@@ -0,0 +1,502 @@
+};
+
+struct nxp_irtc_data {
+	bool is_dst_enabled;
+#ifdef CONFIG_RTC_ALARM
+	rtc_alarm_callback alarm_callback;
+	void *alarm_user_data;
+	uint16_t *alarm_mask;
+	bool is_alarm_pending;
+#endif /* CONFIG_RTC_ALARM */
+#ifdef CONFIG_RTC_UPDATE
+	rtc_update_callback update_callback;
+	void *update_user_data;
+#endif /* CONFIG_RTC_ALARM */
+#ifdef CONFIG_RTC_CALIBRATION
+	int32_t calibration;
+#endif /* CONFIG_RTC_CALIBRATION */
+};
+
+/* The IRTC Offset is 2112 instead of 1990 [2112 - 1990] -> [212] */
+#define RTC_NXP_IRTC_YEAR_OFFSET 212
+
+/* Utility function for testing if bit for alarm mask is enabled. */
+#define RTC_NXP_RTCTIME_COMPARE_FLAG(mask, flag) (((mask) == 0) || (((mask) & (flag)) && (flag)))
+
+/* Utility function to check the bounds of passed in rtc_time */
+static uint8_t is_rtc_time_in_bounds(const struct rtc_time *time_value, uint16_t mask)
+{
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_SECOND)) &&
+	    (time_value->tm_sec < 0 || time_value->tm_sec > 59)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_MINUTE)) &&
+	    (time_value->tm_min < 0 || time_value->tm_min > 59)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_HOUR)) &&
+	    (time_value->tm_hour < 0 || time_value->tm_hour > 23)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_MONTHDAY)) &&
+	    (time_value->tm_mday < 1 || time_value->tm_mday > 31)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_MONTH)) &&
+	    (time_value->tm_mon < 0 || time_value->tm_mon > 11)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_YEAR)) &&
+	    ((time_value->tm_year - RTC_NXP_IRTC_YEAR_OFFSET) < -128 ||
+	     (time_value->tm_year - RTC_NXP_IRTC_YEAR_OFFSET) > 127)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_WEEKDAY)) &&
+	    (time_value->tm_wday < -1 || time_value->tm_wday > 6)) {
+		return 0;
+	}
+
+	if ((mask == 0) && (time_value->tm_isdst < -1 || time_value->tm_isdst > 1)) {
+		return 0;
+	}
+
+	return 1;
+}
+
+/*
+ * The runtime where this is accessed is unknown
+ * so we force a lock on the registers
+ * then force an unlock to guarantee
+ * 2 seconds of write time
+ */
+static void nxp_irtc_unlock_registers(RTC_Type *reg)
+{
+	/* Lock the regsiters */
+	while ((reg->STATUS & (uint16_t)RTC_STATUS_WRITE_PROT_EN_MASK) == 0) {
+		*(uint8_t *)(&reg->STATUS) |= RTC_STATUS_WE(0x2);
+	}
+	/* Unlock the registers */
+	while ((reg->STATUS & (int16_t)RTC_STATUS_WRITE_PROT_EN_MASK) != 0) {
+		/*
+		 * The casting is required to update the
+		 * field of this register, the pattern here
+		 * unlocks the RTC registers for writing.
+		 * When unlocked if a 0x20 if written to this register
+		 * the registers will lock again and the next write
+		 * will lead to a fault.
+		 */
+		*(volatile int8_t *)(&reg->STATUS) = 0x00;
```

Comment:
does this actually need to be casted to uint8_t ? did you try it without casting ? seems to me like it wouldnt be necessary, the msb you can write 0 to I think, since they are w1c iirc (and locked maybe?)

Reply:
The casting is required

Reply:
why?

Reply:
The write to this register need to be a Byte long, otherwise we will create a bus error to trigger.